### DISCUSSION TITLE
How to install ccache when you don't have root access on the machine?

### DISCUSSION THREAD

**Discussion by crazycliffy** [ccache/ccache#641] [Questions & answers] [answered]
I don't have root permissions on my linux machine hence can not install ccache to any root owned folder.  Is it possible copy the ccache archive to a folder (home for instance) and then use it? .  Another benefit of installing it to a shared folder is that I can maintain different versions of ccache.

**Comment by AlexanderLanin** [comment, marked as answer]
Absolutely you can.
There are even numerous options to choose from.

However unfortunately I have not tried these myself. They are theoretical.
I'll try to test them and improve documentation, this is a good hint!
Or you can do it! That would be awesome :-)

I assume you have seen https://github.com/ccache/ccache/blob/master/doc/INSTALL.md
Is there any directory within PATH you have access to or can you change PATH?

Option 1
Install anywhere. Add it to PATH.

Option 2
Assuming your project uses CMake, you can use `-DCMAKE_C_COMPILER_LAUNCHER=/anywhere/ccache -DCMAKE_CXX_COMPILER_LAUNCHER=/anywhere/ccache`.
Or was it `-DRULE_LAUNCH_COMPILE=/anywhere/ccache`? Will edit answer ASAP.

Option 3
Override with e.g. `CC=/anywhere/ccache CXX=/anywhere/ccache cmake ...`

**Comment by jrosdahl** [comment]
@crazycliffy wrote:

>  Is it possible copy the ccache archive to a folder (home for instance) and then use it?

As @AlexanderLanin already mentioned, there are instructions in [the documentation](https://ccache.dev/documentation.html) for your use case. Have you read it? If yes, do you have any suggestion on how to improve it?

**Comment by jrosdahl** [comment]
@AlexanderLanin wrote:
> Option 2
> Assuming your project uses CMake, you can use `-DCMAKE_C_COMPILER_LAUNCHER=/anywhere/ccache -DCMAKE_CXX_COMPILER_LAUNCHER=/anywhere/ccache`.
> Or was it `-DRULE_LAUNCH_COMPILE=/anywhere/ccache`? Will edit answer ASAP.

Both variants work fine.

> Option 3
> Override with e.g. `CC=/anywhere/ccache CXX=/anywhere/ccache cmake ...`

No, that won't work. To be able to set CC and CXX (with CMake) you'll have to use the [symlinks (AKA masquerading) technique](https://ccache.dev/manual/3.7.11.html#_run_modes). For Autotools, it works to use `CC="ccache gcc"`, etc.

**Comment by crazycliffy** [comment]
I am using Option 3 which works ok.  In my Makefile I have something like below...

```
CCACHE_ENV_DIR      = CCACHE_DIR=/any_remote_cache_folder/.cache
CC_ENV              = env PATH=$(anything) LD_LIBRARY_PATH=$(anything) $(CCACHE_ENV_DIR)
CCACHE              = $(ccache_install_folder)/ccache
CC                  = $(CC_ENV) $(CCACHE) $(anywhere)/bin/gcc
CPP                 = $(CC_ENV) $(CCACHE) $(anywhere)/bin/g++
```

So the only downside of the above approach is that anytime I need to use ccache to make queries (checking hit statistics), I'll need to specify where the caching directory is located, so in the meantime I put this alias in my home `.bashrc`  

`alias ccache='env CCACHE_DIR=/any_remote_cache_folder/.cache /ccache_install_folder/ccache'`

Thankfully ccache does not have too many dependencies other than zlib so it's easy to launch without setting up library path.  After compiling ccache, there is a `ccache.1` file that should be installed to the man page database.  Unfortunately, since I don't have root, I can't do a `man ccache`.

**Comment by jrosdahl** [comment]
You can do `man /ccache_install_folder/ccache.1`, though. Or even better, read the HTML documentation. :slightly_smiling_face:
